src/proto/rng.rs

```diff
@@ -0,0 +1,60 @@
+//! `Rng` protocol.
+
+use crate::{data_types::Guid, proto::Protocol, unsafe_guid, Result, Status};
+use core::ptr;
+
+/// Contain a Rng algorithm Guid
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub struct RngAlgorithm(pub Guid);
+
+impl RngAlgorithm {
+    /// Get an empty `RngAlgorithm`
+    ///
+    /// Used provide a buffer to `Rng.get_info`
+    pub fn default() -> Self {
+        Self(Guid::default())
+    }
+}
+
+/// Rng protocol
+#[repr(C)]
+#[unsafe_guid("3152bca5-eade-433d-862e-c01cdc291f44")]
+#[derive(Protocol)]
+pub struct Rng {
+    get_info: extern "efiapi" fn(
+        this: &Rng,
+        algorithm_list_size: *mut usize,
+        algorithm_list: *mut RngAlgorithm,
+    ) -> Status,
+    get_rng: extern "efiapi" fn(
+        this: &Rng,
+        algorithm: *const RngAlgorithm,
+        value_length: usize,
+        value: *mut u8,
+    ) -> Status,
+}
+
+impl Rng {
+    /// Returns information about the random number generation implementation.
+    pub fn get_info(&mut self, algorithm_list: &mut [RngAlgorithm]) -> Result<usize> {
+        let algorithm_list_size = (algorithm_list.len() * 16) as *mut usize;
```

Comment:
I think this is what's causing the crash on aarch64 (not sure why it doesn't crash on the other VMs actually). The `algorithm_list_size` is an in/out variable containing the list size, and you want to pass a pointer to that variable. As currently written though it's casting the list size itself to a pointer.

Try this instead:

```rust
let mut algorithm_list_size = algorithm_list.len() * mem::size_of::<RngAlgorithm>();
```

And then pass `&mut algorithm_list_size` as the second argument to `get_info`.

Reply:
Done